04-Parallel-Simulations-With-MPI/02-Domain-Decomposition.ipynb

```diff
@@ -0,0 +1,446 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In general, **MPI** launches **n** separate **processes** with the given command.\n",
+    "These may or may not be on the same node in a **HPC** cluster, depending on how you request resources in your job script.\n",
+    "Each **process** launched this way is called a **rank** and is given a rank index.\n",
+    "In HOOMD-blue, the **Communicator** class (created by default with **Device**) gives you access to the **rank** information."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\u001b[0;32mimport\u001b[0m \u001b[0mhoomd\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0mdevice\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhoomd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdevice\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCPU\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0mrank\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdevice\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrank\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0mpid\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetpid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf'Hello HOOMD-blue rank {rank} from process id {pid}'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "%pycat hello_hoomd.py"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Hello HOOMD-blue rank 2 from process id 8842\n",
+      "Hello HOOMD-blue rank 3 from process id 8843\n",
+      "Hello HOOMD-blue rank 0 from process id 8840\n",
+      "Hello HOOMD-blue rank 1 from process id 8841\n"
+     ]
+    }
+   ],
+   "source": [
+    "!mpirun -n 4 python3 hello_hoomd.py"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Single program\n",
+    "\n",
+    "HOOMD-blue scripts must be written as a **single program** executed in parallel.\n",
+    "In other words, all **ranks** must load the *same* input file, define the *same* operations with the *same* parameters and triggers, and run the *same* number of time steps.\n",
+    "HOOMD-blue requires this because the operations need to communicate between the ranks.\n",
```

Comment:
```suggestion
    "HOOMD-blue requires this because it decomposes the simulation system into smaller parts and run the same set of operations on each part, and the operations need to communicate between the ranks constantly \n",
```

It wasn't very clear to me why we need to make sure all ranks run the same number of sweeps until I read the domain decomposition section below, so I think it would make more sense to mention it here

Reply:
Good observation. After reading through this again, I decided to move the domain decomposition section to the middle - after describing what ranks are but before going into details on why the single program model is important.